src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs

```diff
@@ -5568,6 +5568,477 @@ await foreach (string value in pets)
             VerifyOperationTreeForTest<BlockSyntax>(source, expectedOperationTree);
         }
 
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/49267")]
+        public void AsyncForeach_StructEnumeratorWithDispose()
+        {
+            var compilation = CreateCompilation(@"
+#pragma warning disable CS1998 // async method lacks awaits
+using System.Threading.Tasks;
+class C
+{
+    static async Task Main()
+    /*<bind>*/{
+        await foreach (var i in new C())
+        {
+        }
+    }/*</bind>*/
+    public AsyncEnumerator GetAsyncEnumerator() => throw null;
+    public struct AsyncEnumerator
+    {
+        public int Current => throw null;
+        public async Task<bool> MoveNextAsync() => throw null;
+        public async ValueTask DisposeAsync() => throw null;
+    }
+}", targetFramework: TargetFramework.NetCoreApp30);
+
+            VerifyOperationTreeAndDiagnosticsForTest<BlockSyntax>(compilation, @"
+IBlockOperation (1 statements) (OperationKind.Block, Type: null) (Syntax: '{ ... }')
+  IForEachLoopOperation (LoopKind.ForEach, IsAsynchronous, Continue Label Id: 0, Exit Label Id: 1) (OperationKind.Loop, Type: null) (Syntax: 'await forea ... }')
+    Locals: Local_1: System.Int32 i
+    LoopControlVariable: 
+      IVariableDeclaratorOperation (Symbol: System.Int32 i) (OperationKind.VariableDeclarator, Type: null) (Syntax: 'var')
+        Initializer: 
+          null
+    Collection: 
+      IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: C, IsImplicit) (Syntax: 'new C()')
+        Conversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+        Operand: 
+          IObjectCreationOperation (Constructor: C..ctor()) (OperationKind.ObjectCreation, Type: C) (Syntax: 'new C()')
+            Arguments(0)
+            Initializer: 
+              null
+    Body: 
+      IBlockOperation (0 statements) (OperationKind.Block, Type: null) (Syntax: '{ ... }')
+    NextVariables(0)
+            ", DiagnosticDescription.None);
+
+            VerifyFlowGraphForTest<BlockSyntax>(compilation, @"");
+        }
+
+        [Fact]
+        public void Foreach_ExtensionGetEnumeratorWithParams()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    /*<bind>*/{
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }/*</bind>*/
+    public sealed class Enumerator
+    {
+        public Enumerator() => throw null;
+        public int Current { get; private set; }
+        public bool MoveNext() => throw null;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this C self, params int[] x) => throw null;
```

Comment:
>params int[] x [](start = 58, length = 14)

Consider testing scenario when this parameter is optional in metadata and doesn't have an array type. Or, for example, has the right type, but it is not the last parameter, there is also another optional parameter on the method.

Reply:
Some error scenarios from source could be interesting too.

---
In reply to: [529094019](https://github.com/dotnet/roslyn/pull/49534#discussion_r529094019) [](ancestors = 529094019)